lib/Optimizer/CodeGen/QuakeCircuitCutPass.cpp

```diff
@@ -0,0 +1,372 @@
+/*******************************************************************************
+ * Copyright (c) 2022 - 2023 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#include "PassDetails.h"
+#include "cudaq/Optimizer/CodeGen/Passes.h"
+#include "cudaq/Optimizer/Dialect/Quake/QuakeOps.h"
+#include "cudaq/Support/GraphPartioner.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+
+using namespace mlir;
+using namespace cudaq;
+
+namespace cudaq::opt {
+#define GEN_PASS_DEF_QUAKECIRCUITCUT
+#include "cudaq/Optimizer/CodeGen/Passes.h.inc"
+} // namespace cudaq::opt
+
+namespace {
+
+/// @brief Given a Quake FuncOp, map it to a Graph.
+void createGraph(func::FuncOp &quakeFunc, Graph &graph) {
+
+  // Track Operation* to the GraphNode representing it
+  std::map<void *, GraphNode> mapper;
+
+  // Track node Ids.
+  std::size_t id = 0;
+
+  // First add |0> qubit nodes to start the graph,
+  // and then add all quantum operations as nodes
+  quakeFunc.walk([&](Operation *op) {
+    if (isa<quake::NullWireOp, quake::OperatorInterface, quake::MxOp,
+            quake::MyOp, quake::MzOp>(op)) {
+      GraphNode node{op, id++};
+      graph.insert({node, std::vector<GraphNode>{}});
+
+      // Also build a map so we can go from Operation* to GraphNode.
+      mapper.insert({op, node});
+    }
+    return WalkResult::advance();
+  });
+
+  // Add the edges now
+  quakeFunc.walk([&](Operation *op) {
+    if (isa<quake::NullWireOp, quake::OperatorInterface>(op)) {
+      auto &thisNode = mapper.at(op);
+      auto users = op->getUsers();
+
+      if (op->hasOneUse()) {
+        // in this case we have an op that produces a qubit,
+        // and that qubit is used as a target.
+        auto otherNode = *users.begin();
+
+        // Get the GraphNode for the user of this op
+        auto &connectedGraphNode = mapper.at(otherNode);
+
+        // connect them i to j
+        graph[thisNode].push_back(connectedGraphNode);
+        return WalkResult::advance();
+      }
+
+      // here we have an op that is ultimately used as a target
+      // but is also used by one or many controls. So we really have one Wire
+      // that we want to thread through the node that is using it as a control,
+      // and connect up the nodes along the way
+      auto lastNode = thisNode;
+
+      // Need to reverse the ordering.
+      // FIXME Check with Eric on this one
+      std::vector<Operation *> opList;
+      for (auto user : users)
+        opList.emplace_back(user);
+      std::reverse(opList.begin(), opList.end());
```

Comment:
The list of users does not provides any guarantees about the order of the operations. For example, after transformations, the op that uses the wire as target might not be the last(or first) in this list. We need to look for the one that uses the wire as target in this list. However, it seems that this code is a bit more complicated than necessary (I might be missing something).

Basically, the value semantics Quake is already a DAG, and thus, when we walk a quake function, we will be visiting operations in topological order. Therefore you only need to look at the current operation, get its operands, and connect them appropriately on the other DAG. We only need to look back to the operations that either defined or used the wire last. Example:

```mlir
%w0 = op ...
%w1 = op2 [%w0] ... // %w0 as control
%w0.1 = op3 %w0     // %w0 as target
```

In this case we need to keep a mapping of which operation defined or last used %w0 . What you could do is:
* After handling `op`, you look a `op2`. Upon seeing its operands, you look for the operations that defined (or used last) them. In this case for `%w0` we get `op`. Then you connect the nodes in the new DAG and update the entry of whom used `%w0` last to `op2`.
* When you look at `op3`, you do the same thing, but at this time the last operation that used `%w0` is `op2` (not `op`).

"Looking back" its easier the "looking forward" in this case (and less error prone---since when looking forward you might run into problem that the operations in the user list might have connections between them but with wires that you are not current looking)

(I'm assuming that we need to convert Quake DAG, which encodes commutative relation on its edges, to a DAG that encodes input/outputs.)

Reply:
This is great feedback, I'll update to simplify.